shell/platform/darwin/ios/framework/Source/FlutterTextInputPlugin.mm

```diff
@@ -1639,22 +1628,28 @@ - (CGRect)firstRectForRange:(UITextRange*)range {
 }
 
 - (CGRect)caretRectForPosition:(UITextPosition*)position {
-  // TODO(cbracken) Implement.
-
-  // As of iOS 14.4, this call is used by iOS's
-  // _UIKeyboardTextSelectionController to determine the position
-  // of the floating cursor when the user force touches the space
-  // bar to initiate floating cursor.
-  //
-  // It is recommended to return a value that's roughly the
-  // center of kSpacePanBounds to make sure the floating cursor
-  // has ample space in all directions and does not hit kSpacePanBounds.
-  // See the comments in beginFloatingCursorAtPoint.
-  return CGRectZero;
-}
-
-- (CGRect)bounds {
-  return _isFloatingCursorActive ? kSpacePanBounds : super.bounds;
+  NSInteger index = ((FlutterTextPosition*)position).index;
+  // Get the bounds of the characters before and after the requested caret position.
+  NSArray<UITextSelectionRect*>* rects =
+      [self selectionRectsForRange:[FlutterTextRange
+                                       rangeWithNSRange:fml::RangeForCharactersInRange(
+                                                            self.text,
+                                                            NSMakeRange(MAX(0, index - 1), 1))]];
+  if (rects.count == 0) {
+    return CGRectZero;
+  }
+  if (index == 0) {
+    // There is no character before the caret, so this will be the bounds of the character after the
+    // caret position.
+    CGRect characterAfterCaret = rects[0].rect;
+    // Return a zero-width rectangle along the left edge of the character after the caret position.
+    return CGRectMake(characterAfterCaret.origin.x, characterAfterCaret.origin.y, 0,
+                      characterAfterCaret.size.height);
+  }
+  CGRect characterBeforeCaret = rects[0].rect;
+  // Return a zero-width rectangle along the right edge of the character before the caret position.
```

Comment:
Maybe use the center instead? I think that would probably work better for RTL and Bidi text since we don't send the bidi info along with the selection rects.

Reply:
I don't think centre would work correctly, it should stay where it is for LTR. But for RTL you're right it likely won't work properly. Maybe a better approach would be putting the rectangle midway between the closest two edges of the adjacent selectionRects to the cursor. Then we don't care about the paint direction v.s. semantic direction.

Reply:
> it should stay where it is for LTR

Sorry I'm not sure how UIKit is going to use this for floating cursor, could you explain why that wouldn't work?

> two edges of the adjacent selectionRects to the cursor.
If the position happens to be at a line break or bidi text boundary then the logically adjacent selection rects won't necessarily be visually adjacent right?